ros/src/tl_detector/tl_detector.py

```diff
@@ -118,72 +100,68 @@ def get_closest_waypoint(self, pose):
+            if stop_wp >= car_pos:  
+                # check for the first light
+                if first:
+                    closest_stop_wp = stop_wp
+                    first = False
+                elif stop_wp < closest_stop_wp:
+                    closest_stop_wp = stop_wp
+        return closest_stop_wp
 
-        Returns:
-            int: index of waypoint closes to the upcoming stop line for a traffic light (-1 if none exists)
-            int: ID of traffic light color (specified in styx_msgs/TrafficLight)
+    def is_tl_visible(self, stop_wp, car_pos):
+        dist_to_light = 10000  
+        if stop_wp and car_pos:
+            dist_to_light = abs(car_pos - stop_wp)
+        return dist_to_light < TL_LOOK_AHEAD
 
+    def process_traffic_lights(self):
+        """Finds closest visible traffic light, 
+            if one exists, and determines its
+            location and color
         """
+        car_pos = None
         closest_light = None
-        closest_light_stop_wp = None
-        dist_to_light = 10000  # initialize to high value
-        first = True
-        stop_line_positions = self.config['stop_line_positions']
-        if(self.pose):
-            car_position = self.get_closest_waypoint(self.pose.pose)
-        else:
-            return -1, TrafficLight.UNKNOWN
-
-        for i, light in enumerate(self.lights):
-            light_stop_pose = Pose()
-            light_stop_pose.position.x = stop_line_positions[i][0]
-            light_stop_pose.position.y = stop_line_positions[i][1]
-            # get the wp closest to each light_position
-            light_stop_wp = self.get_closest_waypoint(light_stop_pose)
-
-            if light_stop_wp >= car_position:  # if waypoint is  close to the traffic light and ahead of the car
-                if first:  # check for the first light
-                    closest_light_stop_wp = light_stop_wp
-                    closest_light = light
-                    first = False
-                elif light_stop_wp < closest_light_stop_wp:
-                    closest_light_stop_wp = light_stop_wp
-                    closest_light = light
-
-        if closest_light_stop_wp is not None:
-            dist_to_light = abs(car_position - closest_light_stop_wp)
-            # we check the status of the traffic light if it's within TL_LOOK_AHEAD waypoints
-            if dist_to_light < TL_LOOK_AHEAD:
-                state = self.get_light_state(closest_light)
-                return closest_light_stop_wp, state
-
-        return -1, TrafficLight.UNKNOWN
+        if self.pose:
+            car_pos = self.get_closest_waypoint(self.pose.pose)
+
+        stop_wp = self.get_stop_waypoint(car_pos)
+        if self.is_tl_visible(stop_wp, car_pos):
+            thread = Thread(
+                    target = self.predict_and_publish, 
+                    args=[stop_wp],
+                    )
+            thread.start()
+
+
+    def predict_and_publish(self, wp):
```

Comment:
Have a few concerns about using the concurrent code in a synchronous way.
Could you please elaborate on why this was chosen over a loop with a fixed rate, as it was done in other modules?

Reply:
Classifier analyzes each image about 0.4 seconds on my computer. And will be faster on GPU. So, rate cannot be fixed